File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int k = 1000000000;
        int l = 1, r = k;
        int res = Integer.MAX_VALUE;
        while (l <= r) {
            int speed = l + (r - l) / 2;
            int time = 0;
            for (int i : piles) {
                time += Math.ceil(i * 1.0 / speed);
            }
            if (time <= h) {
                res = Math.min(res, speed);
                r = speed - 1;
            } else if (time > h) {
                l = speed + 1;
            }
        }
        return res;
    }
}

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        lo = 1
        hi = max(piles)
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if self.verify(piles, mid, h):
                hi = mid - 1
            else:
                lo = mid + 1
        return lo
    def verify(self, piles, speed, h):
        res = 0
        for i in piles:
            res += (math.ceil(i / speed))
        return res <= h